File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string S, string t) {
        unordered_map<char,int>map1;
        unordered_map<char,int>map2;
        int n1=t.size();
        int n2=S.size();
        for(int i=0;i<n1;i++){
            map1[t[i]]++;
        }
        int s=-1,e=-1,len=1e7;
        int i=-1,j=-1;
        int cnt=0;
        while(true){
            bool f1=false;
            bool f2=false;
            while(i<n2-1){
                f1=true;
                i++;
                if(map1.find(S[i])!=map1.end() && map2[S[i]]<map1[S[i]]){
                    cnt++;
                }
                map2[S[i]]++;
                if(cnt==n1){
                    break;
                }
            }

            while(j<i){
                f2=true;
                j++;
                if(cnt==n1){
                    int currlen=i-j+1;
                    if(len>currlen){
                        len=currlen;
                        s=j;
                        e=i;
                    }
                }

                if(map1.find(S[j])!=map1.end()){
                    if(map2[S[j]]<=map1[S[j]]){
                        cnt--;
                    }
                }
                erase(S[j],map2);
                if(cnt<n1){
                    break;
                }

            }
            if(!f1 && !f2)break;
        }
        return len==1e7?"":S.substr(s,e-s+1);
    }

    void erase(char ch,unordered_map<char,int>&map){
        if(map[ch]==1){
            map.erase(ch);
        }
        else{
            map[ch]--;
        }
    }
};

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int n1 = s.length();
        int n2 = t.length();

        int i = -1, j = -1;
        int st = -1, en = -1;
        int ansLen = Integer.MAX_VALUE;
        int len = 0;

        HashMap<Character, Integer> map1 = new HashMap<>();
        HashMap<Character, Integer> map2 = new HashMap<>();

        for (int ind = 0; ind < n2; ind++) {
            map2.compute(t.charAt(ind), (id, val) -> {
                if (val == null)
                    val = 0;
                return val+1;
            });
        }

        while (true) {

            boolean f1 = false;
            boolean f2 = false;

            while (i < n1 - 1) {
                f1 = true;
                i++;
                map1.compute(s.charAt(i), (id, val) -> {
                    if (val == null)
                        val = 0;
                    return val+1;
                });
                if (map2.containsKey(s.charAt(i)) && map1.get(s.charAt(i)) <= map2.get(s.charAt(i))) {
                    len++;
                }
                if (len == n2) {
                    // System.out.println("first-" + i + "," + len);
                    break;
                }
            }

            while (j < i) {
                f2 = true;
                if (len == n2) {
                    // System.out.println("Second" + i + "," + j);
                    int currLen = i - j;
                    if (currLen <= ansLen) {
                        ansLen = currLen;
                        st = j + 1;
                        en = i;
                    }
                }
                j++;
                char ch = s.charAt(j);
                int charcnt;
                if (map1.get(ch) == 1) {
                    map1.remove(ch);
                    charcnt = 0;
                } else {
                    charcnt = map1.compute(ch, (id, val) -> {
                        return val-1;
                    });
                }
                // System.out.println(charcnt + "," + map2.get(ch));

                if (map2.containsKey(ch) && charcnt < map2.get(ch)) {
                    len--;
                }
                if (len < n2)
                    break;
            }

            if (!f1 && !f2)
                break;
        }
        if (ansLen == Integer.MAX_VALUE)
            return "";
        // System.out.println(ansLen);
        return s.substring(st, en + 1);

    }
}